lease/lessor.go

```diff
@@ -642,23 +658,24 @@ func (le *lessor) expireExists() (l *Lease, ok bool, next bool) {
 		return nil, false, false
 	}
 
-	item := le.leaseHeap[0]
+	item := le.leaseHeap.Front()
 	l = le.leaseMap[item.id]
 	if l == nil {
 		// lease has expired or been revoked
 		// no need to revoke (nothing is expiry)
-		heap.Pop(&le.leaseHeap) // O(log N)
+		le.leaseHeap.Pop() // O(log N)
 		return nil, false, true
 	}
-
-	if time.Now().UnixNano() < item.time /* expiration time */ {
+	now := time.Now()
+	if now.UnixNano() < item.time /* expiration time */ {
 		// Candidate expirations are caught up, reinsert this item
 		// and no need to revoke (nothing is expiry)
 		return l, false, false
 	}
-	// if the lease is actually expired, add to the removal list. If it is not expired, we can ignore it because another entry will have been inserted into the heap
 
-	heap.Pop(&le.leaseHeap) // O(log N)
```

Comment:
If we keep this Pop(), i.e., first pop the lease, then push the same lease with a modified time. Do we still need the map inside LeaseHeap?

Reply:
I think we still need, push a modified `leaseWithTime` not garantee a new same lease  `leaseWithTime` produced by `Renew` do not push to the LeaseHeap.

Reply:
So this is fixing a potential bug due to the line 396: https://github.com/etcd-io/etcd/blob/caee28a88e9037a2422eab44c26a5dd11ba8dfa5/lease/lessor.go#L396

Reply:
Can the existing leaseMap be used instead of having a new map?

Reply:
Why not try to find the real reason for revoke failure？

Reply:
@jingyih Yes, we can do it, but i think it's more clear now. what do you think?

Reply:
@yuqitao `revoke` should throught raft log once, so there are some reasons, maybe the network or disk.

Reply:
> @jingyih Yes, we can do it, but i think it's more clear now. what do you think?

I am fine with having a map inside heap. Maybe rename the heap because normally a heap does not behave like this. I am curious if this new bug can be fixed by replacing Pop() with Fix() in line 396?